docs/genai/tutorials/finetune.md

```diff
@@ -0,0 +1,165 @@
+
+## Setup
+
+1. Install Olive
+   
+   This installs Olive from main. Replace with version 0.8.0 when it is released.
+
+   ```bash
+   pip install git+https://github.com/microsoft/olive
+   ```
+
+2. Install ONNX Runtime generate()
+
+   ```
+   pip install onnxruntime-genai
+   ```
+
+3. Install other dependencies
+
+   ```bash
+   pip install optimum peft
+   ```
+
+4. Downgrade torch and transformers
+
+   TODO: There is an export bug with torch 2.5.0 and an incompatibility with transformers>=4.45.0
+
+   ```bash
+   pip uninstall torch
+   pip install torch==2.4
+   pip uninstall transformers
+   pip install transformers==4.44
+   ```
+   
+5. Choose a model
+
+   You can use a model from HuggingFace, or your own model. The model must be a PyTorch model.
+   
+6. Decide whether you are fine-tuning your model, or using a pre-existing adapter
+
+   There are many pre-existing adapters on HuggingFace. If you are using multiple different adapters, these must all use the same fine-tuned layers of the original model.
+
+## Generate model and adapters in ONNX format
+
+1. If fine-tuning, run Olive to fine-tune your model
+
+   Note: this operations requires a system with an NVIDIA GPU, with CUDA installed
+
+   Use the `olive fine-tune` command: https://microsoft.github.io/Olive/features/cli.html#finetune
+
+   Here is an example usage of the command:
+
+   ```bash
+   olive finetune --method qlora -m meta-llama/Meta-Llama-3-8B -d nampdn-ai/tiny-codes --train_split "train[:4096]" --eval_split "train[4096:4224]" --text_template "### Language: {programming_language} \n### Question: {prompt} \n### Answer: {response}" --per_device_train_batch_size 16 --per_device_eval_batch_size 16 --max_steps 150 --logging_steps 50 -o adapters\tiny-codes
+   ```
+
+2. Optionally, quantize your model
+
+   Use the `olive quantize` command: https://microsoft.github.io/Olive/features/cli.html#quantize
+
+
+3. Generate the ONNX model and adapter using the quantized model
+
+   Use the `olive auto-opt` command for this step: https://microsoft.github.io/Olive/features/cli.html#auto-opt
+
+   The `--adapter path` can either be a HuggingFace adapter reference, or a path to the adapter you fine-tuned above.
+
+   The `--provider` argument can be an ONNX Runtime execution provider.
+
+   ```bash
+   olive auto-opt -m <path to your model folder> --adapter_path <path to your adapter> -o <output model folder> --device cpu\|gpu --provider <provider> 
+   ```
+
+4. Convert adapters to `.onnx_adapter` format
+
+   Run this step once for each adapter that you have generated.
+
+   ```bash
+   olive convert-adapters --adapter_path <path to your fine-tuned adapter --output_path <path to .onnx_adapter location --dtype float32
+   ```
+
+## Write your application
+
+This example is shown in Python, but you can also use the C/C++ API, the C# API, and the Java API (_coming soon!_)
+
+```python
+import onnxruntime_genai as og
```

Comment:
While this is ok, shall we update `model-qa.py` in onnxruntime-genai to add support for loading adapters?

Reply:
another excellent idea

Reply:
weird, I thought I commented before - another excellent idea